library/X509/Job.php

```diff
@@ -235,50 +303,68 @@ function (\Exception $exception) use ($target, $streamCapture) {
                 if ($this->finishedTargets % (int) $step == 0) {
                     $this->updateJobStats();
                 }
-                //$loop->stop();
             }
-        )->otherwise(function (\Exception $e) {
-            echo $e->getMessage() . PHP_EOL;
-            echo $e->getTraceAsString() . PHP_EOL;
+        )->otherwise(function (Throwable $e) {
+            Logger::error($e->getMessage());
+            Logger::error($e->getTraceAsString());
         });
     }
 
-    public function getJobId()
+    public function run(): Promise\ExtendedPromiseInterface
     {
-        return $this->jobId;
-    }
+        // Cache job run start time
+        $this->jobRunStart = date("Y-m-d H:i:s");
+        $this->deferred = new Promise\Deferred(function () {
```

Comment:
What about `$promise->always(updateJobStats(true));` Then you can remove the other calls and adjust the method signature.